cmd/bench/internal/client/client.go

```diff
@@ -0,0 +1,250 @@
+type BandwidthClient interface {
+	// Dispatch dispatches a new request, but does not wait for it to complete.
+	Dispatch()
+
+	// Complete waits for the next request to complete. It must wait for whatever
+	// request is the first to complete rather than waiting for a particular
+	// request.
+	Complete()
+}
+
+// RunBandwidthClients runs the provided BandwidthClients for the duration d.
+// For each client, two goroutines are spawned. One loops calling Dispatch, and
+// the other loops calling Complete. The total number of calls to Complete
+// (across all clients) is returned.
+//
+// Note that since only calls to Complete are recorded, it is possible (in fact,
+// quite likely) that there will be outstanding requests (created by Dispatch)
+// when the test quits. This means that the bandwidth estimate is a conservative
+// one. It also means that a very small value for d may skew results by making
+// it so that a large portion of the time is spent waiting for the first request
+// to return (effectively measuring request latency rather than steady state
+// bandwidth).
+func RunBandwidthClients(d time.Duration, clients ...BandwidthClient) uint64 {
+	if len(clients) == 0 {
+		panic("no clients")
+	}
+
+	// the marker will be atomically set to 1 when it's time for the goroutines to
+	// quit
+	var marker uint32
+	stop := func() bool { return atomic.LoadUint32(&marker) == 1 }
+
+	var counts = make([]uint64, len(clients))
+
+	var barrier, wg sync.WaitGroup
+	barrier.Add(1)
+	wg.Add(len(clients))
+	for i, c := range clients {
+		go func(i int, c BandwidthClient) {
+			barrier.Wait()
+			counts[i] = runBandwidthUntil(c, stop)
+			wg.Done()
+		}(i, c)
+	}
+
+	barrier.Done()
+	time.Sleep(d)
+	atomic.StoreUint32(&marker, 1)
+	wg.Wait()
+
+	sum := uint64(0)
+	for _, count := range counts {
+		sum += count
+	}
+	return sum
+}
+
+// runBandwidthUntil spawns two goroutines, with one repeatedly calling
+// c.Dispatch and the other repeatedly calling c.Complete. This continues until
+// stop returns true, at which point the goroutines quit and return the total
+// number of successfully completed requests. Some requests may be left
+// outstanding.
+func runBandwidthUntil(c BandwidthClient, stop func() bool) uint64 {
+	// We don't do any synchronization between the two goroutines, but the
+	// dispatch goroutine will eventually find stop() return false and will quit.
+	go func() {
+		for !stop() {
+			c.Dispatch()
+		}
+	}()
+
+	var count uint64
+	for !stop() {
+		c.Complete()
+		count++
+	}
+	return count
+}
+
+// A LatencyClient is a client used for latency measurements.
+type LatencyClient interface {
+	// Do executes a request and returns the duration.
+	Do() time.Duration
+}
+
+// FuncLatencyClient is a function that implements the LatencyClient interface.
+type FuncLatencyClient func() time.Duration
+
+// Do invokes f.
+func (f FuncLatencyClient) Do() time.Duration { return f() }
+
+var _ LatencyClient = FuncLatencyClient(nil)
```

Comment:
So the code doesn't compile if FuncLatencyClient doesn't implement the LatencyClient interface.